jena-base/src/main/java/org/apache/jena/atlas/lib/Pair.java

```diff
@@ -37,6 +40,30 @@
     public A car() { return a ; }
     public B cdr() { return b ; }
     
+    public static class OfSameType<T> extends Pair<T, T> {
+
```

Comment:
I'm up for another design, as long as `ViewDatasetGraph` and `PairLock` stay reasonably clear and concise. A set of `static` methods could work fine here. I'm not sure what you mean by "even "X extends interface" for the "as boolean" nature?", can you explain that?

Reply:
To be clear, it has to do with `Pair` insofar as it subclasses `Pair`. The idea is to
1. restrict `Pair` to `Pair`s with both elements of the same type and
2. offer some convenience methods that only make sense when both halves of a pair are the same type.

I've broken it out as a separate class, but I'm still up for turning it into a couple of functions. I just don't want to do that until I understand your third point better.

Reply:
Point 3 was the observation 

```
public <S, X> S apply(BiFunction<X, X, S> f, Function<T, X> op) {
         return f.apply(op.apply(a), op.apply(b));
     }
```
could be a function that can acts on `Pair<X,X>`.
```
static public <S, X> S apply(Pair<X,X> pair, BiFunction<X, X, S> f, Function<T, X> op) {
    return f.apply(op.apply(pair.car()), op.apply(pair.cdr()));
}
```
now whether `<S, X>` can meaningfully/usefully be `<S, X extends FOO>` and interface FOO defines the "to boolean" method, 

I don't usually push generics very hard - I leave that to Scala.

Reply:
Okay, I think I see where you are going-- let me try to work this out. Should be able to commit something before the weekend, if we don't get swamped by a blizzard. sigh!

Reply:
It's just a suggestion from looking at the code.  With `OfSameType` a separate class, I'm not worried about the design.  It was coupling what looks to me like specific behaviour to a class that exists primarily because java only returns one thing from a method call.